cajita/src/Cajita_SparseDimPartitioner.hpp

```diff
@@ -0,0 +1,687 @@
+                            int k = static_cast<int>( jnk % rank_k );
+                            current_workload( jnk ) = compute_sub_workload(
+                                di, last_point, point_i, dj, j, dk, k );
+                        } );
+                    // compute the (w_jk^ave - w_jk^{last_point:point_i})
+                    Kokkos::parallel_for(
+                        "compute_diff",
+                        Kokkos::RangePolicy<execution_space>(
+                            0, _ranks_per_dim[dj] * _ranks_per_dim[dk] ),
+                        KOKKOS_LAMBDA( uint32_t jnk ) {
+                            auto wl =
+                                current_workload( jnk ) - ave_workload( jnk );
+                            wl *= wl;
+                            current_workload( jnk ) = wl;
+                        } );
+                    // compute the sum of the difference in all rank_j*rank_k
+                    // regions
+                    int diff;
+                    Kokkos::parallel_reduce(
+                        "diff_reduce",
+                        Kokkos::RangePolicy<execution_space>(
+                            0, _ranks_per_dim[dj] * _ranks_per_dim[dk] ),
+                        KOKKOS_LAMBDA( const int idx, int& update ) {
+                            update += current_workload( idx );
+                        },
+                        diff );
+                    // record the new optimal position
+                    if ( diff <= last_diff )
+                    {
+                        // check if point_i reach the total_tile_num
+                        if ( point_i == rec_mirror( rank, di ) )
+                        {
+                            rec_mirror( current_rank, di ) = point_i;
+                            break;
+                        }
+                        last_diff = diff;
+                        point_i++;
+                    }
+                    else
+                    {
+                        // final optimal position
+                        if ( rec_mirror( current_rank, di ) != point_i - 1 )
+                        {
+                            rec_mirror( current_rank, di ) = point_i - 1;
+                            is_changed = true;
+                        }
+                        last_point = point_i - 1;
+                        break;
+                    }
+                } // end while (optimization for the current rank)
+            }     // end for (all partition/rank in the optimized dimension)
+            Kokkos::deep_copy( _rectangle_partition_dev, rec_mirror );
+        } // end for (3 dimensions)
+    }
+
+    /*!
+      \brief compute the imbalance factor for the current partition
+      \param cart_comm MPI cartesian communicator
+      \return the imbalance factor = workload on current rank / ave_workload
+    */
+    float computeImbalanceFactor( MPI_Comm cart_comm )
```

Comment:
I think we also want `computeImbalance` that just returns the current workload and `computeTotalImbalanceFactor` which sums the imbalance across ranks. That would be a good way to trigger a reoptimization step